tests/shadow/test_shadow_measurement.py

```diff
@@ -0,0 +1,141 @@
+        """Test that the attributes of the MeasurementProcess instance are
+        correctly copied"""
+        dev = qml.device("default.qubit", wires=wires, shots=shots)
+        res = qml.classical_shadow(wires=range(wires), seed_recipes=seed)
+
+        copied_res = copy.copy(res)
+        assert type(copied_res) == type(res)
+        assert copied_res.return_type == res.return_type
+        assert copied_res.wires == res.wires
+        assert copied_res.seed == res.seed
+
+    @pytest.mark.all_interfaces
+    @pytest.mark.parametrize("wires", wires_list)
+    @pytest.mark.parametrize("shots", shots_list)
+    @pytest.mark.parametrize("seed", seed_recipes_list)
+    @pytest.mark.parametrize("interface", ["autograd", "jax", "tf", "torch"])
+    @pytest.mark.parametrize("device", ["default.qubit", "default.mixed", None])
+    def test_format(self, wires, shots, seed, interface, device):
+        """Test that the format of the returned classical shadow
+        measurement is correct"""
+        import torch
+
+        circuit = get_circuit(wires, shots, seed, interface, device)
+        shadow = circuit()
+
+        # test shape and dtype are correct
+        assert shadow.shape == (2, shots, wires)
+        print(shadow.dtype)
+        assert shadow.dtype == np.uint8 if interface != "torch" else torch.uint8
+
+        bits, recipes = shadow
+
+        # test allowed values of bits and recipes
+        assert qml.math.all(np.logical_or(bits == 0, bits == 1))
+        assert qml.math.all(np.logical_or(recipes == 0, np.logical_or(recipes == 1, recipes == 2)))
+
+    @pytest.mark.parametrize("wires", wires_list)
+    @pytest.mark.parametrize("seed", seed_recipes_list)
+    def test_shots_none_error(self, wires, seed):
+        """Test that an error is raised when a device with shots=None is used
+        to obtain classical shadows"""
+        circuit = get_circuit(wires, None, seed)
+
+        msg = "The number of shots has to be explicitly set on the device when using sample-based measurements"
+        with pytest.raises(qml.QuantumFunctionError, match=msg):
+            shadow = circuit()
+
+    @pytest.mark.parametrize("wires", wires_list)
+    @pytest.mark.parametrize("shots", shots_list)
+    def test_multi_measurement_error(self, wires, shots):
+        """Test that an error is raised when classical shadows is returned
+        with other measurement processes"""
+        dev = qml.device("default.qubit", wires=wires, shots=shots)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.Hadamard(wires=0)
+
+            for target in range(1, wires):
+                qml.CNOT(wires=[0, target])
+
+            return qml.classical_shadow(wires=range(wires)), qml.expval(qml.PauliZ(0))
+
+        msg = "Classical shadows cannot be returned in combination with other return types"
+        with pytest.raises(qml.QuantumFunctionError, match=msg):
+            shadow = circuit()
```

Comment:
@eddddddy perhaps I'm missing it, but are there any tests to check that the results being returned are expected? I know this is a hard thing to test, but maybe possible to do by setting a seed?

Reply:
I was thinking that integration tests can be added once the state/expval reconstructions are implemented. I can't think of a way to test the measurement without being able to reconstruct these quantities. What do you think?

Reply:
Yes I agree, I think its best to add integration tests with the second PR. You could add a triviality test, e.g. an empty qnode that outputs the 0state and then checking that all measurement results are as expected within some certainty, e.g.:

```python
dev = qml.device("default.qubit", wires=range(2), shots=100)
@qml.qnode(dev)
def qnode():
    return classical_shadow(wires=range(2))

bitstrings, recipes = qnode()

zero_if_Z = np.mean(bitstrings[np.where(recipes==2)])

assert np.isclose(zero_if_Z, 0, atol=1e-1)
```

Reply:
Ah great idea @Qottmann! I could also test that the bits for `X` and `Y` are roughly uniformly distributed in this case.